### PR TITLE
Fix `onEndReached` bugs for FlatList

### PR DESCRIPTION
**Pull request by ifsnow**
## Summary

Related to : #18887 #16067 #25656 #14015 (There're more issues.)

I'm sure FlatList's `onEndReached` doesn't work as expected. It's one of the headaches that hasn't been solved for quite some time. We had to make it work by modifying `onEndReachedThreshold` several times, but eventually it doesn't work efficiently.

### Problems
1. Unnecessary calls on first render
In most cases, `onEndReached` is called one or more times after the initial rendering.
2. Inefficient calls during scrolling
If scrolling is fast, it will be called again after the `onEndReached` call.
3. Inefficient logic to check
There're some logic implementations that are not needed.

### Solutions
1. Removed unnecessary calls on first render
2. Removed inefficient calls during scrolling
3. Improved and clarified logic to check

## Changelog

[General] [Fixed] - Fix `onEndReached` bugs for FlatList

## Test Plan

[Test App : https://github.com/ifsnow/FlatListImprovementTest](https://github.com/ifsnow/FlatListImprovementTest.git)

- The list has 10 items at the beginning.
- When `onEndReached` is called, 10 items are added.

### 1. After the initial rendering
- Current FlatList

old_flatlist_initial_renering

`OnEndReached` is called twice in a short period and has 30 items.

- Patched FlatList

patched_flatlist_initial_renering

`OnEndReached` is not called and has 10 items.

### 2. After scrolling to the 11th item
- Current FlatList

old_flatlist_scrolling

`OnEndReached` is called twice in a short period, having 50 items.

- Patched FlatList

patched_flatlist_scrolling

`OnEndReached` is called twice when required, having 30 items.

### CODE REVIEW

**Comment by roshangm1** [comment]
This is amazing!!

**Comment by vonovak** [comment]
some automated test would be highly desirable

**Comment by ifsnow** [comment]
@vonovak Thanks for your feedback. I added a test case to check if `onEndReached` is working properly. :)

File: Libraries/Lists/__tests__/VirtualizedList-test.js
Comment (vonovak) [inline Libraries/Lists/__tests__/VirtualizedList-test.js]: I don't understand what this is testing. What does "scrolling back on the screen" mean?

There are a lot of expectations in this test, would it be possible to split it into multiple smaller test cases that would each cover individual cases?

Comment (ifsnow) [reply Libraries/Lists/__tests__/VirtualizedList-test.js]: @vonovak It was a test for this case(scroll down -> scroll to the top -> scroll down again). This test doesn't seem so meaningful. :)

I separated the test because I agree with you. Thanks for your review.

**Comment by ifsnow** [comment]
https://github.com/ifsnow/react-native-infinite-flatlist-patch

Use the patch library above to try it out before this PR merges.

**Comment by tanoabeleyra** [comment]
Thanks!, I'm starting a project with RN and was going to add infinite scrolling today. I didn't know they were so many problems with it since so long ago.

Now I just hope the elevation bug gets fixed :crossed_fingers:

File: Libraries/Lists/__tests__/VirtualizedList-test.js
Comment (vonovak) [inline Libraries/Lists/__tests__/VirtualizedList-test.js]: since all these tests pertain to `onEndReached`, you should wrap them all in `describe('onEndReached', ...)` and remove `onEndReached` from the `it` blocks because it becomes redundant.

File: Libraries/Lists/__tests__/VirtualizedList-test.js
Comment (vonovak) [inline Libraries/Lists/__tests__/VirtualizedList-test.js]: there is quite a bit of duplication in these tests, can you move the duplicated code into the aforementioned `describe` block / beforeEach?

File: Libraries/Lists/__tests__/VirtualizedList-test.js
Comment (vonovak) [inline Libraries/Lists/__tests__/VirtualizedList-test.js]: these functions can live in the describe block

**Comment by vonovak** [comment]
@ifsnow I left some comments that can improve the tests, but please note I'm not a person who can merge the PR, I'm just trying to make it look good.
btw you linked several bugs that you're trying to solve by the PR. If those bugs are reproducible by tests (I don't know) it would be nice to have the tests included (and mention in their description what bugs those tests are covering).

**Comment by ifsnow** [comment]
@vonovak I agree on your comments. I'm sure it can improve the tests, so I modified the test as you said. It's ok even if you don't have merging role. Thank you for your interest in my PR.

File: Libraries/Lists/__tests__/VirtualizedList-test.js
Comment (sahrens) [inline Libraries/Lists/__tests__/VirtualizedList-test.js]: Why wouldn't you want it to be called? The common case is to initially only fetch enough data to fill the viewport, then fetch more to preload the content below the fold so it's immediately available as soon as the user scrolls down to see it. The only case where `onEndReached` should not be called on initial scroll is if you overfetch a bunch of content initially, or you set the `onEndReachedThreshold` very small, neither of which is recommended for optimal performance.

This can be verified by this very test case - if you increase the initial number of items to 30, or drop the `onEndReachedThreshold` to 0.5 (note the default is 2), there will no longer be an initial call to `onEndReached` and this test will pass.

File: Libraries/Lists/__tests__/VirtualizedList-test.js
Comment (sahrens) [inline Libraries/Lists/__tests__/VirtualizedList-test.js]: Same is true here - if you simply reduce `onEndReachedThreshold` to just 1.5 instead of the default of 2 (or tweak how much your test scrolls or renders) this test will pass with no changes to `VirtualizedList`.

**Comment by ifsnow** [comment]
@sahrens Thank you for trying to explain the shortcomings of my implementation.

I didn't think of a situation where the item layout would shrink dynamically. I've tried to improve the logic so that it works more clearly in the various cases you've mentioned. You can see how the improved logic works in the [test app](https://github.com/ifsnow/FlatListImprovementTest).

The first goal is to come up with an easy way to use infinite scrolling in FlatList without worrying about the suitable value of `onEndReachedThreshold` and the number of initial items. 

Please let me know if there is a better approach or something I'm missing.

File: Libraries/Lists/VirtualizedList.js
Comment (ifsnow) [reply Libraries/Lists/VirtualizedList.js]: It was fixed to work properly. Thanks.

File: Libraries/Lists/VirtualizedList.js
Comment (ifsnow) [reply Libraries/Lists/VirtualizedList.js]: You're right, but this is defensive logic to eliminate the possibility of duplicate calls in fast scrolling if `onEndReachedThreshold` is large such as default value(2). You can see the added [test code](https://github.com/facebook/react-native/blob/2d3f6ca9801ef92b446458a2efc795db4ec17021/Libraries/Lists/__tests__/VirtualizedList-test.js#L327).

File: Libraries/Lists/__tests__/VirtualizedList-test.js
Comment (ifsnow) [reply Libraries/Lists/__tests__/VirtualizedList-test.js]: For example, if 20 items are initially loaded
- contentLength = 2000
- visibleLength = 736

The height of all content view(contentLength) is close to three times the height of visible content view(visibleLength). In this case, since there is enough scrolling area, no one would want to load additional data for performance on initial rendering. In the general case, we would try to increase the number of initial data to avoid any performance degradation that may occur with scrolling.

Of course, as you said, we can fix this by adjusting `onEndReachedThreshold` to less than 1. But, finding the best threshold value in each case is difficult and cumbersome. The another issue is that reducing `onEndReachedThreshold` may have side effects at the point where additional data is loaded.

File: Libraries/Lists/__tests__/VirtualizedList-test.js
Comment (ifsnow) [reply Libraries/Lists/__tests__/VirtualizedList-test.js]: The expected behavior is that `onEndReached` will not be called in duplicate while scrolling if we really don't need additional data.

I don't think it's easy to find out the suitable `onEndReachedThreshold` value to match the number and layout of data used. Changing the value each time we use FlatList makes maintenance difficult and increases the chance of mistakes. I'm sure it's better to provide the most common defaults and work in general without any changes.

**Comment by kristerkari** [comment]
This looks very cool! 

Does this PR also solve the issue where`onEndReached` is not getting called consistently between Android and iOS? 

I can test these changes in my project to see if it helps with having to use different `onEndReachedThreshold` value for Android and iOS.

**Comment by ifsnow** [comment]
@kristerkari I think this patch will work the same on both Android and iOS. Please test it on your project and let me know if you have any problems. Thanks.

**Comment by emilioheinz** [comment]
We really need this merged, awesome work.

**Comment by ifsnow** [comment]
@emilioheinz You can use this improvement feature without merged RN. Check it out here https://github.com/ifsnow/react-native-infinite-flatlist-patch

**Comment by luancurti** [comment]
Any updates on this?

**Comment by kristerkari** [comment]
I did a quick test using [patch-package](https://github.com/ds300/patch-package) on top of a 0.61.2 project.

It does not seem to fix the issue with `onEndReachedThreshold: 0` + `onEndReached` working on iOS, but not getting called on Android. Of course there are plenty of more test cases in this issue that could be tested: https://github.com/facebook/react-native/issues/16067#issue-259916910

**Comment by ifsnow** [comment]
@kristerkari Why do you want to set `onEndReachedThreshold` to 0? As you know, default value is 2 and I think it's reasonable to set this value between 1 and 2. Setting it to 0 doesn't seem to be a common case. One of things that we have trouble with using FlatList is that it allows developer to change ambiguous settings.

Please tell me why you need to set it to 0 in general case.

**Comment by kristerkari** [comment]
> Why do you want to set `onEndReachedThreshold` to 0? As you know, default value is 2 and I think it's reasonable to set this value between 1 and 2. Setting it to 0 doesn't seem to be a common case. One of things that we have trouble with using FlatList is that it allows developer to change ambiguous settings.

0 was the value that seemed to be working the best when loading more items using `onEndReached`.

With `onEndReached` being so buggy, I wouldn't really expect people to stick to same values as it clearly does not work consistently.

But anyway it just wanted to point out that I'm using `onEndReachedThreshold: 0` and it works differently on Android and iOS.

**Comment by ifsnow** [comment]
@kristerkari Thanks for your answer. You can see how this patch works differently in this test project(https://github.com/ifsnow/FlatListImprovementTest) I'll check if it can work in various cases.

**Comment by kristerkari** [comment]
@ifsnow It's great that you have a repo for FlatList testing. I'll try it out too.

**Comment by zusinShinpei** [comment]
Anybody here? I hope to merge this.

File: Libraries/Lists/VirtualizedList.js
Comment (Smolyan91) [inline Libraries/Lists/VirtualizedList.js:1406]: where comment ? )

**Comment by ifsnow** [comment]
@zusinShinpei I think it's pretty hard to merge this into a release branch. That's because `FlatList` is one of the components with a lot of history and high impact. So, the improvement of `FlatList` should be carried out by the internal core team, not by the outside.

File: Libraries/Lists/VirtualizedList.js
Comment (Smolyan91) [inline Libraries/Lists/VirtualizedList.js:1402]: magic number 1.5

Comment (fabriziobertoglio1987) [reply Libraries/Lists/VirtualizedList.js:1402]: `_maybeCallOnEndReached` will just `return` if `distanceFromEnd` is `1.5` times higher then then `visibleLength`. I believe currently it is represented with the [`onEndReachedThreeshold`](https://reactnative.dev/docs/flatlist#onendreachedthreshold).

He includes same behaviour in the lines below.

```
const minimumDistanceFromEnd = onEndReachedThreshold * visibleLength;
if (distanceFromEnd >= minimumDistanceFromEnd) {
   return;
}
```

I believe his implementation accepts minimum `onEndReachedThreeshold` to be 1.5 times the `visibleLenght`, as for any value under `1.5 * visibleLength`, the above statement will be triggered

**Comment by seunghyun-woo** [comment]
Great Work. This should really be merged.

**Comment by ghasemikasra39** [comment]
I have the same issue, any solution?

**Comment by fabriziobertoglio1987** [comment]
use the package provided by the author

**Comment by TNBoiBoi** [comment]
Bump, we really need this.

**Comment by vastamaki** [comment]
Any updates on this? Would be nice to get this merged instead of using patch :)

**Review by lunaleaps — CHANGES_REQUESTED** [review]
Thanks @ifsnow for the contribution! 

Indeed this is an issue we've seen in internal usage of FlatList as well. I've been hesitant about pulling this change in because of lack of testing in our list components and any potential side effects this may have in our view logging. We are actively investing in this area but still have a ways to go.

In the meantime, for this fix, could we add an opt-in option to the fix so we can run experimentation around it? That'd help de-risk this change for us as we continue to improve testing and clarify expectations in VL's behavior.

**Comment by KingAmo** [comment]
any news here?

**Comment by dylmye** [comment]
I've been discussing this ticket on the [road to 0.7.0](https://github.com/reactwg/react-native-releases/discussions/26#discussioncomment-3170944) thread, I'd love to pick this up. 

I don't have much in the way of internal React / RN development but I reckon to resolve @lunaleaps request for an opt-in we would need a prop, right?

(Sorry for necro'ing the PR, by the way)

**Comment by lunaleaps** [comment]
@dylmye Sounds great! Let me know if you need any help

**Comment by vksgautam1** [comment]
Any solution of this

**Comment by dylmye** [comment]
I didn't get round to this @vksgautam1 because I don't have write access to ifsnow's repo and (honestly) forgot about this. There's also conflicts that need to be fixed that I don't have the knowledge to fix.

**Comment by fabOnReact** [comment]
https://github.com/ifsnow/react-native-infinite-flatlist-patch/tree/master/patches does not support rn versions > 0.68.
Do you still experience this issue? Should I work on this? Thanks

https://github.com/ifsnow/react-native-infinite-flatlist-patch/issues/11

**Comment by fabOnReact** [comment]
anybody was able to reproduce this issue on 0.72 or 0.73? Thanks

**Comment by fabOnReact** [comment]
@dylmye @vksgautam1 @KingAmo @roshangm1 @tanoabeleyra @luancurti @nnnnnoel @ghasemikasra39 @vastamaki @dylmye @Smolyan91

anybody was able to reproduce this issue on 0.72 or 0.73? Thanks

**Pull request closed without merging**
